### PR TITLE
Make the predicated accumulator the split-K form AutoTileMatmulL0 emits and accepts

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary

A split-K accumulator written as a peeled branch — a fresh `tile.matmul` seed on
the first K step, `tile.matmul_acc` afterwards — gives one logical value two
producers on two L0C buffers. Nothing reads L0C except the FIXPIPE drain, whose
destinations are L1 and GM, so no `Acc`->`Acc` copy exists to reconcile the arms:
the shape is one the machine cannot represent. `tile.matmul_acc`'s optional
`init_cond` operand models the MAD's `cmatrixInit` bit directly and expresses the
same reduction on a single buffer.

`init_cond` has shipped since #2403, but the compiler both generated the peeled
shape and refused the predicated one, so the operand was unusable in practice.
This makes `AutoTileMatmulL0` emit it and accept it.

## Changes

- **`BuildMatmulBody` emits a predicated accumulate** for a plain `tile.matmul`
  K-loop instead of an `if ko == 0` phi over a fresh matmul and an in-place
  accumulate. `tile.matmul_bias` keeps a head-peel — it has no `init_cond`
  operand and its first K step must apply the bias exactly once.
- **`AnalyzeMatmul` accepts arity 4** for the accumulate kind. A caller-written
  predicate previously opted the call out of L0 tiling entirely. The predicate is
  threaded to all three emission sites in `BuildKLoopRewrite`:

  | Site | `init_cond` |
  | --- | --- |
  | Pipelined K-loop (`num_full >= 2`) | `user_cond and ko == 0` |
  | Lone straight-line full block (`num_full == 1`) | `user_cond` verbatim |
  | Partial tail | none — it is never the first K block |

  Forwarding the predicate to the tail would re-zero the accumulator and discard
  the full blocks' partial sum.
- **`MatchCanonicalSplitKAcc` grows a second arm** for the predicated shape, so a
  kernel written with `init_cond` still gets the canonical split-K retile and
  therefore M/N tiling. Without it, migrating a kernel would be a silent perf
  cliff: `TryMNFold` skips `is_acc()`, so the canonical pre-phase is the only
  route to M/N tiling for a split-K accumulator. `matmul`/`phi` become nullable
  and the retiler falls back to the accumulate call's result var.

Two supporting fixes found along the way:

- **`python_printer`** prints `tensor.matmul_acc`'s `init_cond` as a keyword. The
  Tensor-level DSL signature is `(acc, lhs, rhs, a_trans, b_trans, init_cond)`, so
  positional slot 4 is `a_trans`; a positional predicate re-parsed as a transpose
  flag and then collided with the printed `a_trans=` kwarg.
- **The emitter's literal-predicate fold** matched only `ConstInt`. Simplify's
  arithmetic rewriter produces `ConstBool`, so after `LowerPipelineLoops` the fold
  stopped firing and a replicated K-loop emitted a branch on a constant.

## Backward compatibility

**No user-facing API changes** — nothing under `python/` or any `.pyi`.

`MemoryReuse` keeps repairing hand-peeled accumulators. An earlier revision of
this work replaced that repair with a hard diagnostic; that broke previously
compiling kernels (e.g. a hand-peeled 176x192x176 stage-2 K-loop), so the repair
is retained. It now carries a comment recording that the predicated form is the
direction of travel and that retiring the repair needs the source migration to
land first — pypto-lib alone has 95 peeled sites, and they all still compile.

## Effect on generated code

Measured on the `gemm.py` shape from pypto-lib's examples, compiled both ways
through the Default pipeline:

| | peeled | predicated |
| --- | --- | --- |
| L0C (`loc=acc`) buffers | 3 | **1** |
| L0A / L0B buffers | 2 / 2 | **1 / 1** |
| `pto.tmatmul` + `.acc` | 2 | 2 |
| `scf.if` | 1 | 1 |

Instruction count is a wash — codegen still expands a runtime predicate into
`scf.if { pto.tmatmul } else { pto.tmatmul.acc }`, since ptoas exposes no single
op carrying the init bit. The win is that one logical value can no longer land on
two buffers, which matters at production tile sizes where a single accumulator
already fills L0C.

## Verification

Rebased onto `main` at `a5113628`, rebuilt, and re-run:

- `tests/ut`: **10304 passed, 3 skipped, 2 xfailed**, no failures.
- `pre-commit` on all changed files: 20 hooks pass, including
  `check-docs-en-zh-parity`, `check-emitter-check-classification`,
  `check-op-name-literals`, `cpplint`, `ruff`, and `pyright`.
- pypto-lib downstream sweep: 52 kernel scripts using the peeled idiom (95 sites)
  compiled against this work, each in its own process — no new diagnostics. The
  one failure is a pypto-lib-side `AttributeError` at import, unrelated.
- Not run: `tests/st`, and any on-device execution.

An earlier revision of this branch also carried the `YieldFixupMutator`
`As<Var>` -> `AsVarLike` fix. #2484 landed that independently, so it dropped out
of this diff on rebase and is no longer part of this PR.

## Reviewer notes

- The predicate composition is the part most worth a careful read. Walk
  `num_full >= 2` + tail, `num_full == 1` + tail, and the `allow_k_boundary`
  case — an off-by-one there is silent wrong numerics.
- New tests for the two K-boundary cases use the `Before`/`Expected` +
  `assert_structural_equal` pattern; both were mutation-checked (adding a
  spurious predicate to the tail in `Expected` fails the test).
- `tests/ut/ir/transforms/conftest.py` pins **Ascend950**, which decides the L0
  tile the chooser picks. Hand-writing an `Expected` under a different backend
  yields a different tiling.

### CODE REVIEW

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 1]
```diff
@@ -297,6 +297,33 @@ class TopDownRetargeter {
...
+  /// the peel must keep compiling, and pypto-lib alone has 95 such sites.
+  ///
+  /// Retiring this repair is therefore a *long-horizon* intent, not a near-term
+  /// plan.  It would need the source-level migration to land first, and it is
+  /// worth doing eventually because the repair is not sound in general: the
+  /// branch-exclusivity argument below does not cover a post-if read of the
+  /// accumulator that bypasses the phi, where the seed path's data is returned
+  /// with no diagnostic at any layer.  Until then, prefer `init_cond` in new
+  /// code and leave existing peels alone.
+  ///
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/memory_reuse_pass.cpp:326]: Not changing this here — the defect is real, but it is pre-existing and this PR deliberately leaves the repair untouched.

`git diff upstream/main..HEAD -- src/ir/transforms/memory_reuse_pass.cpp` is **comment-only**: zero non-comment lines. The unsound retarget you describe is exactly what the comment you are quoting documents ("does not cover a post-if read of the accumulator that bypasses the phi, where the seed path's data is returned with no diagnostic at any layer"). The comment is the change; the logic is untouched upstream code.

An earlier revision of this branch *did* remove this repair and replace it with a hard `ValueError` naming both allocations. It was reverted, because it broke previously-compiling kernels — e.g. a hand-peeled 176x192x176 stage-2 K-loop that had compiled fine. Retaining backward compatibility for the peeled form was a deliberate call for this PR.

Doing it properly is the third option — keep the repair, but add the soundness test the current code lacks (a post-`IfStmt` read of the accumulator that bypasses the phi), repairing when provably safe and reporting only when not. That needs its own dominance/liveness analysis and regression tests, and it changes behaviour for shapes it declines, so it belongs in its own PR rather than bolted onto this one. Worth noting the blast radius is bounded: `AutoTileMatmulL0` no longer *generates* this shape after this PR, so every remaining path into it is a peel in author source.

Leaving this thread open for maintainer judgement on whether to schedule that follow-up.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by lyfne123 — APPROVED** [review]
